server: support using the ApiListener from a worker thread

xavidotron opened the pull request:
Commit Message: server: support using the ApiListener from a worker thread
Additional Description:

Before this change, the SyntheticReadCallbacks used by the ApiListener hard-coded the main thread dispatcher, resulting in dispatcher assertion failures if the ApiListener was used from a worker thread. This change allows it to use the appropriate dispatcher whether used from the main thread or a worker thread.

This change is motivated by a project at Google that needs to run synthetic requests through the Envoy filter chain at a relatively high QPS; doing so via the ApiListener seems much cleaner and more efficient than doing so via a loopback interface. Thus, to sustain a higher level of QPS efficiently, we seek to use the ApiListener from multiple worker threads in parallel.

Risk Level: Low.
Testing: test/server:api_listener_test test/integration:api_listener_integration_test
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

RyanTheOptimist commented:
/assign @alyssawilk

alyssawilk commented:
so I'm fine with allowing non-main-thread dispatchers but this seems a roundabout way of doing so.  Why not simply pass a Dispatcher in the consturctor? 
/wait-any

xavidotron commented:
Actually, thinking about it a bit more, the wrapper object from my option 1 could itself implement the Http::ApiListener interface, which makes that seem cleaner than I initially thought, so I'm thinking of giving that a try unless you prefer otherwise.

alyssawilk commented:
SG let's give that a shot and see how it looks
/wait

xavidotron commented:
Here's a new version taking that approach. It involved some Envoy Mobile changes to accommodate the new ownership model, but seems pretty reasonable to me.

alyssawilk commented on mobile/library/common/http/client.h:
```diff
@@ -51,9 +51,9 @@ struct HttpClientStats {
 class Client : public Logger::Loggable<Logger::Id::http> {
 public:
-  Client(ApiListener& api_listener, Event::ProvisionalDispatcher& dispatcher, Stats::Scope& scope,
+  Client(ApiListenerPtr api_listener, Event::ProvisionalDispatcher& dispatcher, Stats::Scope& scope,
```
&& for ownership clarity?

xavidotron replied:
Done.

alyssawilk commented on source/server/api_listener_impl.cc:
```diff
@@ -31,6 +31,33 @@ void ApiListenerImplBase::SyntheticReadCallbacks::SyntheticConnection::raiseConn
 
+// This class wraps a HttpConnectionManager used as an Http::ApiListener and the associated
+// SyntheticReadCallbacks to ensure that both objects have the same lifetime.
+class HttpApiListener::HttpConnectionManagerWrapper : public Http::ApiListener {
```
I'd be inclined to call it an APIListenerWrapper, since it's wrapping the listener first and the HCM implicitly but I'll let @RyanTheOptimist tie break.

xavidotron replied:
Yeah, I thought that Alyssa's suggestion made sense so I went ahead and renamed the wrapper class.

alyssawilk commented on test/server/api_listener_test.cc:
```diff
@@ -163,21 +163,26 @@ name: test_api_listener
   ASSERT_EQ("test_api_listener", http_api_listener.name());
   ASSERT_EQ(ApiListener::Type::HttpApiListener, http_api_listener.type());
-  ASSERT_TRUE(http_api_listener.http().has_value());
+  auto api_listener = http_api_listener.createHttpApiListener(server_.dispatcher());
```
Can we make the wrapper class public and remove the outer readCallbacksForTest with a dynamic cast to the wrapper class here?

xavidotron replied:
Done.

alyssawilk commented on source/server/api_listener_impl.cc:
```diff
@@ -31,6 +31,33 @@ void ApiListenerImplBase::SyntheticReadCallbacks::SyntheticConnection::raiseConn
+    return http_connection_manager_->newStream(response_encoder, is_internally_created);
+  }
+
+  SyntheticReadCallbacks& readCallbacksForTest() { return read_callbacks_; }
```
this doesn't have hacky dynamic casts so I think we could make it a regular accessor and remove the forTests label

xavidotron replied:
Done.

alyssawilk reviewed (commented):
yeah I like this much better, thanks for the refactor
/retest and adding @RyanTheOptimist as another reviewer

RyanTheOptimist commented:
Are these failures real, or flakes?
```
[  FAILED  ] IpVersionsAndGrpcTypes/ListenerMultiAddressesIntegrationTest.MultipleLdsUpdatesSharingListenSocketFactoryWithMultiAddresses/1, where GetParam() = (4-byte object <00-00 00-00>, 4-byte object <01-00 00-00>)
[  FAILED  ] IpVersionsAndGrpcTypes/ListenerMultiAddressesIntegrationTest.MultipleLdsUpdatesSharingListenSocketFactoryWithMultiAddresses/2, where GetParam() = (4-byte object <01-00 00-00>, 4-byte object <00-00 00-00>)
```

xavidotron commented:
Seems like unrelated timeouts, let me retry them.

/retest

alyssawilk commented:
your call but I think we can land this as is given the code is E-M only.  Then again it's located in core code so either way.

RyanTheOptimist commented:
Yeah, given that it's in core I'm inclined to err on the side of caution but I do agree that it's safe.

mattklein123 commented:
Is there some context behind this change? Is this for some in-process use case? Mostly just trying to understand so someone doesn't come along and delete this later. Can you at minimum add some motivation to the PR description or possible add some comments to the interfaces you are changing? Thank you.

/wait-any

xavidotron commented:
Added something to the PR description and added an explanation of the need for the dispatcher parameter to the ApiListener interface; does this look good?

mattklein123 reviewed (approved):
Thanks!

xavidotron commented:
/retest

xavidotron commented:
/retest
